File: java-practice/spring-data-jpa-study/src/main/java/org/jpa/chap01/api/PostApi.java
package org.jpa.chap01.api;

import lombok.RequiredArgsConstructor;
import org.jpa.chap01.dto.PostRequest;
import org.jpa.chap01.entity.PostEntity;
import org.jpa.chap01.repository.PostJpaRepository;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/posts")
@RequiredArgsConstructor
public class PostApi {
    private final PostJpaRepository postJpaRepository;

    @GetMapping
    public List<PostEntity> getPostList() {
        return postJpaRepository.findAll();
    }

    @PostMapping
    public PostEntity createPost(
            @RequestBody final PostRequest.Create request
    ) {
        return postJpaRepository.save(request.toEntity());
    }

    @DeleteMapping
    public String deletePost(
            @RequestBody final PostRequest.Delete request
    ) {
        postJpaRepository.deleteById(request.postId());
        return "deleted";
    }

    @PutMapping
    public PostEntity updatePost(
            @RequestBody final PostRequest.Update request
    ) {
        Optional<PostEntity> byId = postJpaRepository.findById(request.postId());

        if (byId.isPresent()) {
            PostEntity postEntity = byId.get();
            postEntity.updateTitle(request.title());
            postEntity.updateContent(request.content());
            return postEntity;
        }

        return null;
    }
}

File: api-practice/reactive-cache/reactive-cache-api/src/main/kotlin/dev/lkdcode/domain/api/PostApi.kt
package dev.lkdcode.domain.api

import dev.lkdcode.domain.service.PostService
import dev.lkdcode.domain.value.Post
import org.springframework.web.bind.annotation.*
import reactor.core.scheduler.Schedulers

@RestController
class PostApi(
    private val postService: PostService,
) {

    @GetMapping("/posts/{postId}")
    fun fetch(@PathVariable(name = "postId") postId: Long): MonoResponseEntity<Post> =
//        postService
//            .fetch(postId)
//            .map { monoResponseEntity(it) }
        MonoResponseEntity.fromCallable { postService.fetch(postId) }
            .subscribeOn(Schedulers.boundedElastic())
            .flatMap { it }
            .map { monoResponseEntity(it) }

    @PostMapping("/posts/{postId}")
    fun create(@PathVariable(name = "postId") postId: Long): MonoResponseEntity<Post> =
        postService
            .create(postId)
            .map { monoResponseEntity(it) }

    @PutMapping("/posts/{postId}")
    fun update(@PathVariable(name = "postId") postId: Long): MonoResponseEntity<Post> =
        postService
            .update(postId)
            .map { monoResponseEntity(it) }

    @DeleteMapping("/posts/{postId}")
    fun delete(@PathVariable(name = "postId") postId: Long): MonoResponseEntity<Post> =
        postService
            .delete(postId)
            .map { monoResponseEntity(it) }
}